undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterAdd.java

```diff
@@ -54,12 +53,20 @@ class FilterAdd extends AbstractAddStepHandler {
     @Override
     protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
         final String name = context.getCurrentAddressValue();
-        final ServiceTarget target = context.getServiceTarget();
-        final ServiceBuilder<?> sb = target.addService(UndertowService.FILTER.append(name));
-        final Consumer<HandlerWrapper> serviceConsumer = sb.provides(UndertowService.FILTER.append(name));
+        CapabilityServiceBuilder<?> csb = context.getCapabilityServiceTarget().addCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition());
+        Consumer<HandlerWrapper> serviceConsumer =  csb.provides(FilterCapabilities.FILTER_CAPABILITY.getDefinition(), UndertowService.FILTER.append(name));
         HandlerWrapper wrapper = this.factory.createHandlerWrapper(context, model);
-        sb.setInstance(Service.newInstance(serviceConsumer, wrapper));
-        sb.setInitialMode(ServiceController.Mode.ON_DEMAND);
-        sb.install();
+        csb.setInstance(new FilterProviderService(serviceConsumer, wrapper));
+        csb.setInitialMode(ServiceController.Mode.ON_DEMAND);
+        csb.install();
     }
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+            throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+        //NOTE: its dynamic by design.
+        context.registerCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition().fromBaseCapability(context.getCurrentAddress()));
+    }
+
```

Comment:
You do not need to override this method.  Capability registration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractAddStepHandler.java#L216

Reply:
AFAIR Brian said it has to be done this way.

Reply:
Provided that the filter's ResourceDefinition registers the RuntimeCapability (and it should), the capability will be auto-registered with the OperationContext by the super implementation.

Reply:
@baranowb @pferraro My Sept 7, 2023 comment about using this method was in reference to FilterRefAdd, not FilterAdd.

The issue is this:

https://github.com/wildfly/wildfly/pull/17134/files#diff-8a27bf46cb96032958c3cb8549c72bb98a745b73e2f3aa2756eaeb698f532a1fR130

```
final Supplier<HandlerWrapper> fSupplier = csb.requiresCapability(Capabilities.CAPABILITY_FILTER, HandlerWrapper.class, name);
```

This is a capability requirement that is not enforced at Stage.MODEL; it's only enforced at Stage.RUNTIME via this service dependency. So an admin-only server can be used to create an invalid config that's only detected when the config is switched out of admin-only. And the bad config doesn't abort boot, which it should.

The issue is this capability reference is not done via an attribute; it's the filter-ref resource's own name that specifies the name of the required filter. So it's at the level of the filter-ref resource that we need to record the capability requirement in Stage.MODEL. AbstractAddStepHandler. recordCapabilitiesAndRequirements is the place I could think of where such a requirement can be registered.